### PR TITLE
(Feature) integrate server with zoom

### PR DESCRIPTION
**Pull request by serdiukov-o-nordwhale**
# Description 

[x] add /:provider param to face verification endpoint
[x] add Zoom enrollment API
[x] add EnrollmentProcessor as service responsible to perform verifications
[x] move all session / wallet updates to the EnrollmentProcessor
[x] add Zoom/Kairos enrollment providers, perform verification using corresponding APIs (for kairos use existing code)

About #159

# How Has This Been Tested?

see https://github.com/GoodDollar/GoodDAPP/pull/1551

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [x] I have added screenshots related to my pull request ( for frontend tasks)
- [x] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: src/server/storage/storageAPI.js [hunk 1]
```diff
@@ -126,22 +122,9 @@ const setup = (app: Router, storage: StorageAPI) => {
     '/user/delete',
     passport.authenticate('jwt', { session: false }),
     wrapAsync(async (req, res, next) => {
-      const { user, log, body } = req
-      const { zoomSignature, zoomId } = body
+      const { user, log } = req
       log.info('delete user', { user })
 
-      if (zoomId && zoomSignature) {
```
Comment (sirpy) [inline src/server/storage/storageAPI.js:133]: why was this removed?
in client code we still need to use our special zoomId as id when enrolling the user

File: src/server/server.config.js [hunk 2]
```diff
@@ -232,24 +246,18 @@ const conf = convict({
...
-  zoomToken: {
-    doc: 'Zoom APP Token',
+  zoomLicenseKey: {
+    doc: 'Zoom (Face Recognition / Liveness Test API) License key',
     format: '*',
-    env: 'ZOOM_TOKEN',
+    env: 'ZOOM_LICENSE_KEY',
     default: ''
   },
-  zoomMinMatchLevel: {
```
Comment (sirpy) [inline src/server/server.config.js:247]: still need this.
you removed the logic of search

File: src/server/verification/processor/EnrollmentProcessor.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  async enroll(payload: any, providerType: EnrollmentProvider = EnrollmentProviders.Kairos) {
+    const { sessionId } = payload
+    const sessionRef = GunDBPublic.session(sessionId)
+    const providerInstance = this._createProvider(providerType)
+
+    this.sessionRef = sessionRef
+
+    try {
+      const enrollmentResult = await providerInstance.enroll(payload, this.user.identifier)
```
Comment (sirpy) [inline src/server/verification/processor/EnrollmentProcessor.js:47]: 1. we first search if highly similar face exists (where we use the min match level env variable)
if highly similar face exists then we return failure
2. we use zoomId a special identifier sent by the user from the client and not this.user.identifier
3. if identifier is already enrolled we need to catch this possibility and not return failure, but user already enrolled

File: src/server/verification/processor/EnrollmentProcessor.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    try {
+      await Promise.all([
+        adminApi.whitelistUser(gdAddress, profilePublickey),
+        storage.updateUser({ identifier: loggedInAs, isVerified: true })
+      ])
+
+      sessionRef.put({ isWhitelisted: true })
+    } catch ({ message }) {
+      sessionRef.put({ isWhitelisted: false, isError: message })
```
Comment (sirpy) [inline src/server/verification/processor/EnrollmentProcessor.js:94]: needs to return error to the client, are we using gundb session in new zoom client integration?

File: src/server/verification/processor/EnrollmentProcessor.js (hunk 3 above)
Comment (sirpy) [reply src/server/verification/processor/EnrollmentProcessor.js:47]: see old zoom logic and kairos addIfUniqueAndAlive logic

File: src/server/server-middlewares.js [hunk 5]
```diff
@@ -39,4 +41,8 @@ export default (app: Router, env: any) => {
     log.error(error)
     res.status(400).json({ message: error.message })
   })
+
+  // register & start cron tasks
```
Comment (serdiukov-o-nordwhale) [inline src/server/server-middlewares.js:45]: forgot that DisposeEnrollmentsTask returns factory 
so you should do:

```
import createDisposeEnrollmentsTask from './verification/cron/DisposeEnrollmentsTask'


const disposeEnrollmentsTask = createDisposeEnrollmentsTask(UserDBPrivate)

CronTasksRunner.registerTask(disposeEnrollmentsTask)
CronTasksRunner.startTasks()
```

Comment (serdiukov-o-nordwhale) [reply src/server/server-middlewares.js:45]: Also we should handle express app termination (or process exit / SIGINT / SIGTERM event) and call CronTasksRunner.stopTasks()

**Comment by serdiukov-o-nordwhale** [comment]
```
Snapshot Summary
4654 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `npm run coverage -- -u` to update them.
```

File: src/server/verification/processor/__tests__/EnrollmentProcessor.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    invokeMap([updateUserMock, updateSessionMock, GunDBPublic.session, AdminWallet.whitelistUser], 'mockReset')
+
+    zoomServiceMock.reset()
+  })
+
+  afterAll(() => {
+    modulesToMock.forEach(jest.unmock)
+
+    zoomServiceMock.restore()
+    zoomServiceMock = null
```
Comment (serdiukov-o-nordwhale) [inline src/server/verification/processor/__tests__/EnrollmentProcessor.js:54]: add enrollmentProcessor = null

File: src/server/verification/processor/provider/ZoomProvider.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      ({ enrollmentIdentifier: matchId }) => matchId.toLowerCase() !== enrollmentIdentifier.toLowerCase()
+    )
+    // if there're at least one record left - we have a duplicate
+    const isDuplicate = !!duplicate
+
+    // notifying about duplicates found or not
+    await notifyProcessor({ isDuplicate })
+
+    if (duplicate) {
+      const duplicateFoundMessage = `Duplicate with identifier '${duplicate.enrollmentIdentifier}' found.`
```
Comment (sirpy) [inline src/server/verification/processor/provider/ZoomProvider.js:72]: make sure this isnt returned to user

File: src/server/verification/processor/provider/ZoomProvider.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+      // rethrowing unexpected errors (e.g. no conneciton or service error)
+      if (!response) {
+        throw exception
+      }
+
+      // if exception isn't related to the case when
+      // facemap is just already enrolled
+      if ('nameCollision' !== response.subCode) {
+        const isEnrolled = false
```
Comment (sirpy) [inline src/server/verification/processor/provider/ZoomProvider.js:95]: please follow same logic for already enrolled as in previous implementation
or explain better in comments how is already enrolled is passed from the thrown exception

File: src/server/verification/processor/provider/ZoomProvider.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    const throwCustomException = (errorMessage, errorResponse) => {
+      const exception = new Error(errorMessage)
+
+      exception.response = { isVerified: false, ...errorResponse }
+      throw exception
+    }
+
+    // 1. checking liveness
+    try {
+      await api.detectLiveness(payload)
```
Comment (sirpy) [inline src/server/verification/processor/provider/ZoomProvider.js:37]: we dont need this. we get liveness result from enroll

File: src/server/verification/verificationAPI.js [hunk 10]
```diff
@@ -13,96 +12,87 @@ import { sendOTP, generateOTP } from '../../imports/otp'
...
+      const { enrollmentIdentifier } = params
+
+      let enrollmentResult
+      const processor = createEnrollmentProcessor(storage)
+
+      try {
+        processor.validate(user, enrollmentIdentifier, payload)
+
+        if (user.isVerified || conf.skipFaceVerification) {
+          const sessionRef = GunDBPublic.session(sessionId)
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js:79]: if user.isVerified we need to make sure to whitelist him, maybe we changed the whitelisting contract and he is no longer whitelisted there, so we trust that we already whitelistedhim in the past and whitelist him again in the new contract

File: src/server/verification/processor/EnrollmentProcessor.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      // TODO: enqueue enrollmentIdentifier to the corresponding mongo collection using storage
+      // add current timestamp to each collection item
+    }
+  }
+
+  async disposeEnqueuedEnrollments(onProcessed: (identifier: string, exception?: Error) => void): Promise<void> {
+    noop(onProcessed) // eslint / lgtm stub
+    // TODO
+
+    // 1. get all items from enqueued enrollment identifiers collection
```
Comment (sirpy) [inline src/server/verification/processor/EnrollmentProcessor.js:85]: we could be running multiple nodes of GoodServer, so this needs to happen atomically (so two servers dont try to delete same record)
it can be done either by using mongo findOneAndUpdate or use the queueMongo lock that we use to synchronize blockchain calls across servers

**Comment by serdiukov-o-nordwhale** [comment]
Merged, all changes required described in the #166

File: src/server/server-middlewares.js (hunk 5 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/server-middlewares.js]: @serdiukov-o-nordwhale 
the changes with DisposeEnrollmentsTask Stas already did.
I added changes for process exit / SIGINT / SIGTERM events here - https://github.com/GoodDollar/GoodServer/pull/177
